probes/internal/utils/permissions/permissions.go

```diff
@@ -0,0 +1,311 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package permissions
+
+import (
+	"embed"
+	"fmt"
+	"strings"
+
+	"github.com/ossf/scorecard/v4/checker"
+	sce "github.com/ossf/scorecard/v4/errors"
+	"github.com/ossf/scorecard/v4/finding"
+	"github.com/ossf/scorecard/v4/probes/internal/utils/uerror"
+)
+
+func createText(t checker.TokenPermission) (string, error) {
+	// By default, use the message already present.
+	if t.Msg != nil {
+		return *t.Msg, nil
+	}
+
+	// Ensure there's no implementation bug.
+	if t.LocationType == nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, "locationType is nil")
+	}
+
+	// Use a different text depending on the type.
+	if t.Type == checker.PermissionLevelUndeclared {
+		return fmt.Sprintf("no %s permission defined", *t.LocationType), nil
+	}
+
+	if t.Value == nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, "Value fields is nil")
+	}
+
+	if t.Name == nil {
+		return fmt.Sprintf("%s permissions set to '%v'", *t.LocationType,
+			*t.Value), nil
+	}
+
+	return fmt.Sprintf("%s '%v' permission set to '%v'", *t.LocationType,
+		*t.Name, *t.Value), nil
+}
+
+func CreateNegativeFinding(r checker.TokenPermission,
+	probe string,
+	fs embed.FS,
+) (*finding.Finding, error) {
+	// Create finding
+	text, err := createText(r)
+	if err != nil {
+		return nil, fmt.Errorf("create finding: %w", err)
+	}
+	f, err := finding.NewWith(fs, probe,
+		text, nil, finding.OutcomeNegative)
+	if err != nil {
+		return nil, fmt.Errorf("create finding: %w", err)
+	}
+
+	// Create Location
+	var loc *finding.Location
+	if r.File != nil {
+		loc = &finding.Location{
+			Type:      r.File.Type,
+			Path:      r.File.Path,
+			LineStart: newUint(r.File.Offset),
+		}
+		if r.File.Snippet != "" {
+			loc.Snippet = newStr(r.File.Snippet)
+		}
+		f = f.WithLocation(loc)
+		f = f.WithRemediationMetadata(map[string]string{
+			"repo":     r.Remediation.Repo,
+			"branch":   r.Remediation.Branch,
+			"workflow": strings.TrimPrefix(f.Location.Path, ".github/workflows/"),
+		})
+	}
+	if r.LocationType != nil {
+		f = f.WithValue("permissionLocation", string(*r.LocationType))
+	}
+	if r.Name != nil {
+		f = f.WithValue("tokenName", *r.Name)
+	}
+	f = f.WithValue("permissionLevel", string(r.Type))
+	return f, nil
+}
+
+func CreateFindings(fs embed.FS,
+	raw *checker.RawResults,
+	locationType checker.PermissionLocation,
+	permissionLevel checker.PermissionLevel,
+	probe, tokenName string,
+) ([]finding.Finding, string, error) {
+	if raw == nil {
+		return nil, "", fmt.Errorf("%w: raw", uerror.ErrNil)
+	}
+
+	results := raw.TokenPermissionsResults
+	var findings []finding.Finding
+
+	if results.NumTokens == 0 {
+		f, err := finding.NewWith(fs, probe,
+			"No token permissions found",
+			nil, finding.OutcomeNotAvailable)
```

Comment:
notapplicable?

Reply:
changed in https://github.com/ossf/scorecard/pull/3816/commits/5eed9362c9b73d5554740fa836b5d78cbd2f59c5